nixos/modules/services/networking/syncthing.nix

```diff
@@ -216,7 +216,9 @@ in {
               default = name;
               description = lib.mdDoc ''
                 The path to the folder which should be shared.
-              '';
+                If it is not an absolute path, it is relative to
+                [dataDir](#opt-services.syncthing.dataDir).
+              '';  # because the working directory is set to `dataDir`
```

Comment:
Yeah this also seems like a breaking change to me. Whereas previously `path = "foo"` would be `/foo`, it's not `/var/lib/syncthing/foo` instead.

But also, are relative paths even useful? How about just documenting that people can use `path = config.services.syncthing.dataDir + "/some/path"` if one wants it relative to the syncthing data directory. We could even disallow relative paths entirely to get rid of the potential confusion ("relative to what?").

Reply:
> Yeah this also seems like a breaking change to me.

It is a breaking change but I doubt that anyone who wanted to use `/foo` used a relative path. I could add a changelog entry though.

> But also, are relative paths even useful?

They are useful insofar as it means `path` is set to a sane default. I e.g. have a server where I don't care what path is actually used as long as it's one under `/var/lib/syncthing`.

Reply:
> It is a breaking change but I doubt that anyone who wanted to use `/foo` used a relative path

The problem is we don't know. We can't assume nobody uses a feature, even if it's a misfeature. Even if we don't get any reports from users about it breaking, there could be a lot of users we never ever hear from that get frustrated.

Note however that breaking changes are only frustrating if they're silent. If we properly deprecate something, we can do breaking changes smoothly. This should always be preferred over silent breakages whenever possible.

In this case, if we wanted to require `path` to be absolute it would look like this:
- Warn when a relative path is passed, instructing the user to convert it to an absolute path by adding a `/` in front
- Wait at least one release, after which the warning can be turned into an error
- Wait at least one release, ideally more, before the error message is removed. At that point we may introduce relative paths again.

Assuming people update to every release, nobody will have a bad time like that.

Of course, another way of specifying the relative path could also be introduced if that's worthwhile, like a `relativePath` option, which could be used in the transition period of the above. This would be overkill here imo, but it's good to think about these things.

Reply:
Just realized that an interesting outtake of having a huge userbase with known configs is that we can query statistics on feature usage.

Reply:
> The problem is we don't know. We can't assume nobody uses a feature, even if it's a misfeature. Even if we don't get any reports from users about it breaking, there could be a lot of users we never ever hear from that get frustrated.

I think we all agree that the likelihood of anyone using this is negligable. That's why I think a changelog entry should be enough.

> Note however that breaking changes are only frustrating if they're silent. If we properly deprecate something, we can do breaking changes smoothly. This should always be preferred over silent breakages whenever possible.

The change wouldn't happen silently. We'd put it in the breaking changes section of the changelog. Its existence shows that breaking changes do happen once in a while and people know they are supposed to read the changelog.

Reply:
> I think we all agree that the likelihood of anyone using this is negligable.

Never underestimate the amount of hacks people are willing to deal with. Yes, this might be a case where it's safe, but who knows? Unless we have some clear evidence that actually nobody is using it, we should not make such assumptions.

Honestly I don't care about this example, but I've been bitten by so many small backwards incompatible changes to NixOS in the past that broke my setup silently (not just NixOS, upstream as well). Due to the way NixOS works, we often have the capability to deprecate things properly, so whenever possible we should go for that and make it the norm. I want to live in a world where I can update my NixOS system and be 100% sure that everything that worked before continues working.

> The change wouldn't happen silently. We'd put it in the breaking changes section of the changelog.

That is effectively a silent breakage because nobody reads through all the release notes and makes sure their config isn't affected by it.

Reply:
After thinking about it more, we should really be more careful with this PR. Syncthing is a distributed service, and debugging those is extremely painful. I've messed around with Syncthing lately, believe me, you don't want to debug it. I would really prefer it if we used a softer transition.

> In this case, if we wanted to require `path` to be absolute it would look like this:
> 
> * Warn when a relative path is passed, instructing the user to convert it to an absolute path by adding a `/` in front
> * Wait at least one release, after which the warning can be turned into an error
> * Wait at least one release, ideally more, before the error message is removed. At that point we may introduce relative paths again.

I think this is a good suggestion. We can do the following:

-  If the path is absolute, leave it as is
- If the path begins with a '~/`, leave it as is (it will be relative to the `dataDir`
- If it's neither (for example, `path = "foo";`), convert it to absolute path (`path = "/foo";`) and throw a warning
- Document the behavior above
- Write tests for all three cases

That way, we aren't doing any breaking changes. After one release we can turn the warning into an error. After one more release we introduce relative paths. We can also do it sooner or later, it really depends.

There's also a good reason to do that: Syncthing doesn't actually have any paths that look like `foo/bar`; it only has paths `/foo/bar` and `~/foo/bar`. We aren't solving a Syncthing problem here. We are solving a minor UX problem for NixOS. It is important to not break configs in order to achieve better UX, since having breaking changes is not really a good UX, **ESPECIALLY** if they are silent.

Reply:
> If it's neither (for example, path = "foo";), convert it to absolute path (path = "/foo";) and throw a warning

If the goal is to avoid any silent breakage whatsoever, we also shouldn't do this. A user might have overriden `systemd.services.syncthing.serviceConfig.WorkingDirectory` to a more sensible value. If we're going to do a transition, ambiguous configurations should result in errors right away.

Reply:
Fair point. I thought that override to be unlikely, but yes, that is a possibility.

We can throw an error outright, then, and leave a comment in code instruction to remove the error in 23.05 and to make paths relative by default. How does that sound?

Reply:
I don't want to write code that judges whether a path is relative. I suggest using `types.path` and removing the default value. This would throw an error for anyone using a relative path or one starting with `~`. For 23.05, the default can be set to `cfg.dataDir + "/" + name`.

Reply:
> I don't want to write code that judges whether a path is relative

Neither do I, lmao. But I think the whole procedure boils down to checking if the string starts with `~/`, and then checking if the string minus the starting `~` is a valid `types.path`. If it turns out to be too hard, then I'm fine with ditching the `~` idea.

> I suggest using `types.path` and removing the default value. This would throw an error for anyone using a relative path or one starting with `~`. For 23.05, the default can be set to `cfg.dataDir + "/" + name`.

Then we wouldn't be supporting all of the upstream semantics for no reason other than "we didn't care enough". It's fine: care is a limited resource, but we also would also create some edge cases like this:
```nix
"nested-directory" = {
  path = "/var/lib/syncthing/parent-directory/nested-directory";
  # or path = "${services.syncthing.dataDir}/parent-directory/nested-directory";
};
```
While we could otherwise use this:
```nix
"nested-directory" = {
  path = "~/parent-directory/nested-directory";
};
```
Don't get me wrong, this is not a HUGE deal. But we might as well do our best while we're at it, right?